---
title: "Metro Wastewater COVID-19 Monitoring"
author: "Metropolitan Council"
date: "January 6, 2023"
output:
  html_document:
    css: ["www/rmd-report-style.css", "www/colors.css", "www/font.css"]
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
setwd(here::here())
library(ggplot2)
library(shiny)
library(tidyverse)
library(lubridate)
library(plotly)
library(councilR) # remotes::install_github("Metropolitan-Council/councilR", force = T)
library(DT)
library(gh)


case_data <- read.csv("data/case_data.csv") %>%
  mutate(date = as.Date(date)) %>%
  mutate(across(where(is.numeric), round, digits = 4)) 

load_data <- read.csv("data/clean_load_data.csv") %>%
  mutate(date = as.Date(date)) %>%
  left_join(case_data) %>%
  mutate(across(where(is.numeric), round, digits = 4))

variant_data <- read.csv("data/clean_variant_data.csv") %>%
  mutate(date = as.Date(date)) %>%
  mutate(across(where(is.numeric), round, digits = 2)) %>%
  filter( !variant %in% c("Omicron BA.4 and BA.5")) %>%
  mutate(variant = factor(
    variant,
    levels = c(
      "Alpha, Beta & Gamma",
      "Delta",
      "Omicron BA.1",
      "Omicron BA.2.12.1",
      "Omicron BA.4 and BA.5",
      "Omicron BA.4",
      "Omicron BA.5",
      "Omicron BA.5 (Excluding BQ.1)",
      "Omicron BQ.1",
      "Omicron BA.2 (Excluding BA.2.12.1)"
    )))
  
copies_by_variant <- read.csv("data/copies_by_variant.csv") %>%
    mutate(variant = factor(
    variant,
    levels = c(
      "Alpha, Beta & Gamma",
      "Delta",
      "Omicron BA.1",
      "Omicron BA.2.12.1",
      "Omicron BA.4 and BA.5",
      "Omicron BA.4",
      "Omicron BA.5",
      "Omicron BA.5 (Excluding BQ.1)",
      "Omicron BQ.1",
      "Omicron BA.2 (Excluding BA.2.12.1)"
    )
  )) %>%
  mutate(date = as.Date(date)) %>%
  mutate(across(where(is.numeric), round, digits = 2))  %>%
  filter(!(variant == "Omicron BA.4 and BA.5" &
             date > "2022-05-30")) %>%
  filter(date > "2021-04-01")


omi_ratio_data <- read.csv("data/omi_ratio_data.csv") %>%
  mutate(date = as.Date(date)) %>%
  mutate(across(where(is.numeric), round, digits = 3))

font_family_list <- "Roman, Helvetica, Tahoma, Geneva, Arial, sans-serif"

whiteSmoke <- "#F5F5F5"

pal <- c("Total Viral Load" = "white",
         "Omicron BA.2 (Excluding BA.2.12.1)" = "#D64776",
         "Omicron BQ.1" = "#006400",
         "Omicron BA.5 (Excluding BQ.1)" = "#000080",
         "Omicron BA.4" = "#3D9F93",
         "Omicron BA.4 and BA.5" = "#A9A3FE",
         "Omicron BA.2.12.1" = "#FBC740",
         "Omicron BA.1" = "#6D3571",
         "Delta" = "#1D94B7",
         "Alpha, Beta & Gamma" = "#84BB25"
         )

ann_list <- list(
  text = paste(
    "<br><br>",
    "<i>", "Last sample date",
    max(c(
      load_data$date,
      variant_data$date
      # case_data$date,
      # combined_data$date
    ), na.rm = T),
    "</i>"
  ),
  font = list(
    size = 11,
    family = font_family_list,
    color = councilR::colors$suppBlack
  ),
  x = 1,
  y = -0.2,
  showarrow = F,
  xref = "paper", yref = "paper",
  xanchor = "right", yanchor = "auto",
  xshift = 0, yshift = -25
)

hov_lab_list <- list(
  font = list(
    size = 20,
    family = font_family_list,
    color = councilR::colors$suppWhite
  ),
  # bgcolor = "white",
  stroke = list(
    councilR::colors$suppGray,
    councilR::colors$suppGray,
    councilR::colors$suppGray,
    councilR::colors$suppGray
  ),
  padding = list(l = 5, r = 5, b = 5, t = 5)
)
```

## Total Viral Load
#### Tracking COVID-19 Prevalence with Metro Plant Wastewater

The daily amount of SARS-CoV-2 viral RNA flowing into the Metro Plant correlates closely with the number of new daily cases reported by the Minnesota Department of Health. The plant serves nearly 2 million people in Minneapolis, Saint Paul, and 64 other metro area communities.

###### How to read this graph:
The blue line and points show the total amount of SARS-CoV-2 viral RNA in wastewater flowing into the Metro Plant, in millions copies of the SARS-CoV-2 genome per person served by the wastewater plant, per day. Blue points are daily values; the blue line is a running average of the previous 7 days. The gray line shows the number of new daily COVID-19 cases in the Metro Plant's service area per 100,000 residents, by sample collection date (data from the Minnesota Department of Health). The most recent case data (darker gray) are incomplete and subject to change.

Last Sample Date: `r max(load_data$date, na.rm = T)`
```{r load plot, out.width = '100%', message = FALSE, warning = FALSE}
whiteSmoke <- "#F5F5F5"

hov_lab_list <- list(
  font = list(
    size = 20,
    family = "Arial Narrow",
    color = councilR::colors$suppWhite
  ),
  # bgcolor = "white",
  stroke = list(
    councilR::colors$suppGray,
    councilR::colors$suppGray,
    councilR::colors$suppGray,
    councilR::colors$suppGray
  ),
  padding = list(l = 5, r = 5, b = 5, t = 5)
)

left_axis_text <- list(
  tickfont = list(color = colors$councilBlue,
                  size = 16,
                  family = "Arial Narrow"),
  overlaying = "y",
  side = "left",
  zerolinewidth = 0,
  zerolinecolor = colors$suppWhite,
  gridcolor = colors$suppWhite,
  rangemode = "nonnegative"
)

left_axis_title <- list(
  x = -0.05,
  y = 1.05,
  text = "Viral load in wastewater<br>M copies/person/day",
  xref = "paper",
  yref = "paper",
  showarrow = F,
  align = "left",
  font = list(
    size = 16,
    family = "Arial Narrow",
    color = councilR::colors$councilBlue
  )
)

right_axis_title <- list(
  x = 1,
  y = 1.05,
  text = "New Daily COVID-19 cases<br>per 100K residents",
  xref = "paper",
  yref = "paper",
  showarrow = F,
  align = "right",
  font = list(
    size = 16,
    family = "Arial Narrow",
    color = councilR::colors$suppBlack
  )
)


load_plot <-
  load_data %>%
  # left_join(case_data, by = "date") %>%
  plot_ly(type = "scatter", mode = "lines",
          width = 900, height = 500) %>%
  add_trace(
    mode = "markers",
    x = ~date,
    y = ~copies_day_person_M_mn,
    name = "Viral load, Metro Plant service area",
    size = 1,
    yaxis = "y2",
    marker = list(
      color = "rgba(0, 84, 164, .5)",
      size = 8,
      line = list(
        color = colors$councilBlue,
        width = 0.5
      )
    ),
    # fillcolor = ,
    # line = list(width = 2, color = colors$esBlue),
    hoverinfo = "text",
    text = ~hover_text_load
  ) %>%
  add_trace(
    mode = "lines",
    x = ~date,
    y = ~copies_day_person_7day,
    name = "7-day avg. viral load",
    size = 1,
    yaxis = "y2",
    # fill = "tozeroy",
    # fillcolor = "rgba(0, 154, 199, .5)",
    line = list(width = 2, color = colors$councilBlue),
    hoverinfo = "text",
    text = ~hover_text_load_7day
  ) %>%
  add_trace(
    x = ~slice(load_data, 1:(n()-7))$date,
    y = ~slice(load_data, 1:(n()-7))$covid_cases_7day,
    name = "7-day avg. cases per capita",
    fill = "tozeroy",
    fillcolor = "rgba(160, 160, 160, .3)",
    line = list(width = 0.5, color = colors$suppGray),
    hoverinfo = "text",
    text = ~slice(load_data, 1:(n()-7))$hover_text_case
  ) %>%
  add_trace(
    x = ~slice(load_data, (n() - 6):n())$date,
    y = ~slice(load_data, (n() - 6):n())$covid_cases_7day,
    name = "7-day avg. cases per capita, Incomplete",
    fill = "tozeroy",
    fillcolor = "rgba(160, 160, 160, .8)",
    line = list(width = 0.5, color = colors$suppGray),
    hoverinfo = "text",
    text = ~slice(load_data, (n() - 6):n())$hover_text_case
  ) %>%
  layout(
    annotations = list(left_axis_title, right_axis_title),
    autosize = T,
    showlegend = TRUE,
    margin = list(l = 75, r = 75, b = 75, pad = 10),
    hovermode = "closest",
    hoverdistance = "10",
    hoverlabel = hov_lab_list,
    yaxis2 = left_axis_text,
    xaxis = list(
      title = list(
        range = ~c(min(date)-4, max(date) + 4),
        text = "",
        standoff = 0,
        font = list(
          size = 16,
          family = "Arial Narrow",
          color = councilR::colors$suppBlack
        )
      ),
      zerolinewidth = 2,
      gridcolor = colors$suppWhite,
      zerolinecolor = colors$suppWhite,
      tickfont = list(
        size = 16,
        family = "Arial Narrow",
        color = councilR::colors$suppBlack
      )
    ),
    yaxis = list(
      side = "right",
      title = list(
        standoff = 25,
        font = list(
          size = 16,
          family = "Arial Narrow",
          color = councilR::colors$suppBlack
        )
      ),
      zerolinewidth = 1,
      tickfont = list(
        size = 16,
        family = "Arial Narrow",
        color = councilR::colors$suppBlack
      ),
      gridcolor = colors$suppWhite,
      zerolinecolor = colors$suppWhite
    ),
    legend = list(
      orientation = "h",
      font = list(
        size = 16,
        family = "Arial Narrow",
        color = councilR::colors$suppBlack
      )
    )
  ) %>%
  config(displayModeBar = F) 

div(load_plot, align = "center")
```

\newpage
## COVID-19 Variants
The dominance of various SARS-CoV-2 variants in wastewater entering the Metro Plant in Saint Paul has ebbed and flowed over time. Omicron BA.1 replaced Delta as the dominant variant in late December 2021. In February 2022, we observed the appearance and increase of Omicron BA.2. In April, sub-lineage BA.2.12.1 increased in prevalence, becoming the dominant variant by mid-May.  In July, subvariant BA.5 took over as the dominant SARS-CoV-2 variant in Metro Plant influent wastewater. BQ.1 is currently the most prevalent BA.5 subvariant we observe.

###### How to read this graph:
This graph shows the frequency of SARS-CoV-2 variants in Metro Plant wastewater. This frequency is the percentage of the total viral RNA that was identified as being from each particular variant.  Points are daily data; lines and shaded areas are averages of the previous 7 days.

```{r variant plot, message = FALSE, warning = FALSE}

variant_freq_title <- list(
            x = -0.05,
            y = 1.05,
            text = "Variant frequency (%)<br>indicated by maker genes",
            xref = "paper",
            yref = "paper",
            showarrow = F,
            align = "left",
            font = list(
              size = 14,
              family = font_family_list,
              color = councilR::colors$suppBlack
            )
          )

variant_plot <- 
variant_data %>%
      plot_ly(# set height and width
              width = 900, height = 500) %>%
      add_trace(
        type = "scatter",
        mode = "lines",
        x = ~date,
        fill = "tozeroy",
        y = ~frequency_7day,
        split = ~variant,
        color = ~variant,
        alpha = 0.25,
        colors = pal,
        hoverinfo = "none",
        showlegend = F
      ) %>%
        add_trace(
        type = "scatter",
        mode = "markers",
        x = ~date,
        y = ~frequency,
        split = ~variant,
        color = ~variant,
        alpha = 0.8,
        colors = pal,
        hoverinfo = "text",
        text = ~hover_text_variant
      ) %>%
      layout(
        annotations = list(ann_list, variant_freq_title),
        hovermode = "closest",
        hoverdistance = "10",
        hoverlabel = hov_lab_list,
        margin = list(
          l = 50,
          r = 100,
          b = 50,
          pad = 10
        ),
        xaxis = list(
          title = list(
            text = "", standoff = 25,
            font = list(
              size = 14,
              family = font_family_list,
              color = councilR::colors$suppBlack
            )
          ),
          zerolinewidth = 2,
          zeroline = TRUE,
          showline = FALSE,
          showgrid = FALSE,
          tickfont = list(
            size = 12,
            family = font_family_list,
            color = councilR::colors$suppBlack
          )
        ),
        yaxis = list(
          title = list(
           # text = "<b>Frequency of marker genes (%)</b>",
            standoff = 25,
            font = list(
              size = 14,
              family = font_family_list,
              color = councilR::colors$suppBlack
            )
          ),
          tickformat = "1%",
          tickfont = list(
            size = 12,
            family = font_family_list,
            color = councilR::colors$suppBlack
          ),
          gridcolor = "gray90",
          zerolinecolor = "gray50",
          zerolinewidth = 2,
          range = c(0, 1.1)
        ),
        legend = list(
          orientation = "h",
          font = list(
            size = 11,
            family = font_family_list,
            color = councilR::colors$suppBlack
          )
        )
      ) %>%
      config(displayModeBar = FALSE)

div(variant_plot, align = "center")
```

\newpage
## COVID-19 Load by Variant
#### Another way of looking at the variant data
This plot shows the total amount (“load”) of each variant entering the Metro Plant, over time.  BA.5 has been the dominant variant since early July 2022. A BA.5 subvariant, BQ.1, is currently replacing other BA.5 sub-lineages, but the total prevalence of BA.5 remains above 75%. Simultaneously, however, the BA.2 subvariants, BA.2.75 and XBB, are increasing in prevalence.

###### How to read this graph:
This graph shows the measured number of copies per person per day of each variant in wastewater. Shaded areas are averages of the previous 7 days for each variant load. The gray line is the 7-day average total viral load. Variant proportions do not always add to 100%, so slight discrepancies between the total viral load and that of the individual variants are expected. 

Last Sample Date: `r max(load_data$date, na.rm = T)`
```{r load x variant plot, message = FALSE, warning = FALSE}

variant_prev_title <- list(
  x = -0.05,
  y = 1.15,
  text = "Viral load in wastewater<br>M copies_variant/person/day",
  xref = "paper",
  yref = "paper",
  showarrow = F,
  align = "left",
  font = list(
    size = 14,
    family = "Arial Narrow",
    color = councilR::colors$suppBlack
  )
)

date_anno2 <- list(
  text = paste(
    "<br><br>",
    "<i>", "Latest sample date",
    max(c(
      load_data$date,
      copies_by_variant$date
      # case_data$date,
      # combined_data$date
    ), na.rm = T),
    "</i>"
  ),
  font = list(
    size = 11,
    family = font_family_list,
    color = councilR::colors$suppBlack
  ),
  x = 1,
  y = -0.35,
  showarrow = F,
  xref = "paper", yref = "paper",
  xanchor = "right", yanchor = "auto",
  xshift = 0, yshift = -25
)


base_ggplot <-
  copies_by_variant %>%
  rename(`7-day-average` = hover_text_variant_7day) %>%
  mutate(variant = factor(variant, levels = rev(levels(variant)))) %>%
  ggplot(aes(x = date, y = copies_7day)) +
  geom_area(
    position = "stack",
    aes(
      x = date,
      y = copies_7day,
      color = variant,
      fill = variant,
      label = `7-day-average`
    ),
    alpha = 0.75,
    na.rm = T,
    lwd = 0.25
  )  +
  # solid black line - total covid load
  geom_line(
    data = load_data[load_data$date > min(copies_by_variant$date),] %>%
      mutate(variant = "Total COVID-19 Load") %>%
      rename(`7-day-average` = hover_text_load_7day),
    aes(
      x = date,
      y = copies_day_person_7day,
      color = variant,
      fill = variant,
      label = `7-day-average`
    ),
    lwd = 0.6
  ) +
  scale_fill_manual(values = pal,
                    drop = T) + 
  scale_color_manual(values = pal,
                     drop = T) +
  scale_y_continuous(name = "",
                     labels = scales::unit_format(unit = "M")) +
  scale_x_date(name = "Date",
               breaks = "2 months",
               date_labels = "%b '%y") +
  geom_hline(yintercept = 0, color = "gray30", size = 0.5) + 
  theme_void()+
  theme(panel.grid.major.x= element_blank(),
        axis.line = element_blank(),
        axis.text.x = element_text(),
        axis.text.y = element_text())

variant_load_plot <- 
ggplotly(base_ggplot, tooltip = c("label"),
         width = 900, height = 500) %>%
  layout(
    hoverlabel = list(
      font = list(
        size = 20,
        family = "Arial Narrow",
        color = councilR::colors$suppWhite
      ),
      # bgcolor = "white",
      stroke = list(
        councilR::colors$suppGray,
        councilR::colors$suppGray,
        councilR::colors$suppGray,
        councilR::colors$suppGray
      ),
      padding = list(l = 5, r = 5, b = 5, t = 5)
    ),
    margin = list(l = 125, pad = 0),
    legend = list(
      orientation = "h",
      yanchor = "top",
      xanchor = "left",
      title = list(text = ""),
      font = list(
        size = 16,
        family = "Arial Narrow",
        color = councilR::colors$suppBlack
      )),
    yaxis = list(
      title = list(text = ""),
      tickfont = list(
        size = 16,
        family = "Arial Narrow",
        color = councilR::colors$suppBlack
      ),
      rangemode = "nonnegative",
      gridcolor = "gray90",
      zerolinecolor = "gray90",
      zerolinewidth = 0
    ),
    xaxis = list(
      title = list(text = ""),
      zeroline = FALSE,
      showline = FALSE,
      showgrid = FALSE,
      tickfont = list(
        size = 16,
        family = "Arial Narrow",
        color = councilR::colors$suppBlack
      )),
    annotations = list(
      list(
        x = -0.16,
        y = 0.6,
        text = "Viral load<br>(Million copies<br>per person,<br>per day)",
        xref = "paper",
        yref = "paper",
        showarrow = F,
        borderpad = 0,
        borderwidth = 0,
        align = "center",
        font = list(
          size = 16,
          family = "Arial Narrow",
          color = councilR::colors$suppBlack
        ))
    )
  )


div(variant_load_plot, align = "center")
```
\newpage
###### More information about variant detection
Variant frequencies are inferred from the presence of key mutations in the SARS-CoV-2 genome. Alpha, Beta and Gamma frequencies are inferred from the presence of the N501Y mutation; Delta from the L452R mutation; and Omicron from the K417N mutation. Some variants share mutations: presence of K417N mutation before November 18 were inferred to be the Beta variant. The two sub-lineages of Omicron (BA.1 and BA.2) are distinguished by the HV 69/70 deletion: Omicron BA.1 contains both the K417N mutation and the HV 69/70 deletion. Omicron BA.2 has the K417N mutation but not the HV 69/70 deletion.  Omicron BA.2.12.1 has the L452Q mutation.  BA.4 and BA.5 marked the return of the L452R mutation, which was seen previously in Delta. BA.4 and BA.5 are measured using the ORF7b_L11F and M_D3N mutation assays, respectively. BQ.1 is a subvariant derived from BA.5 and is measured here using the N_E136D mutation assay. This assay measures all BA.5.3.1 sub-lineages, but BQ.1 sub-lineages currently make up greater than 90% of locally-circulating BA.5.3.1 sub-lineages, and we report the assay results here as “BQ.1.”
